### ISSUE TITLE
Generated SpecificRecord classes with Lists fail when reading using Avro's SpecificDatumReader

### ISSUE THREAD

**Issue by tackley** [julianpeeters/avrohugger#25]
I have an avro schema that includes a list field. The generated SpecificRecord avrohugger code generated includes a get implementation as follows:

``` scala
  def get(field: Int): AnyRef = {
    field match {
      case pos if pos == 0 => {
        java.util.Arrays.asList(({
          tests map { x =>
            x
          }
        }: _*))
      }.asInstanceOf[AnyRef]
      case _ => new org.apache.avro.AvroRuntimeException("Bad index")
    }
  }
```

`org.apache.avro.specific.SpecificDatumReader` explicitly attempts to re-use the specific record objects when reading subsequent records, and for List fields it attempts to call `.clear()` on the list it read on the previous object. The list returned by `java.util.Arrays.asList` does not support `.clear`, as shown by the following stack trace of a spark job:

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0, localhost): java.lang.UnsupportedOperationException
    at java.util.AbstractList.remove(AbstractList.java:161)
    at java.util.AbstractList$Itr.remove(AbstractList.java:374)
    at java.util.AbstractList.removeRange(AbstractList.java:571)
    at java.util.AbstractList.clear(AbstractList.java:234)
    at org.apache.avro.generic.GenericDatumReader.newArray(GenericDatumReader.java:330)
    at org.apache.avro.reflect.ReflectDatumReader.newArray(ReflectDatumReader.java:77)
    at org.apache.avro.reflect.ReflectDatumReader.readArray(ReflectDatumReader.java:119)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:153)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:155)
    at org.apache.avro.generic.GenericDatumReader.readField(GenericDatumReader.java:193)
    at org.apache.avro.reflect.ReflectDatumReader.readField(ReflectDatumReader.java:230)
    at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:183)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:151)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:155)
    at org.apache.avro.generic.GenericDatumReader.readField(GenericDatumReader.java:193)
    at org.apache.avro.reflect.ReflectDatumReader.readField(ReflectDatumReader.java:230)
    at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:183)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:151)
    at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:142)
    at org.apache.avro.file.DataFileStream.next(DataFileStream.java:233)
    at org.apache.avro.mapred.AvroRecordReader.next(AvroRecordReader.java:66)
    at org.apache.avro.mapred.AvroRecordReader.next(AvroRecordReader.java:32)
    at org.apache.spark.rdd.HadoopRDD$$anon$1.getNext(HadoopRDD.scala:248)
    at org.apache.spark.rdd.HadoopRDD$$anon$1.getNext(HadoopRDD.scala:216)
    at org.apache.spark.util.NextIterator.hasNext(NextIterator.scala:71)
    at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
    at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
    at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
    at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
    at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply$mcV$sp(PairRDDFunctions.scala:1108)
    at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply(PairRDDFunctions.scala:1108)
    at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply(PairRDDFunctions.scala:1108)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1285)
    at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13.apply(PairRDDFunctions.scala:1116)
    at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13.apply(PairRDDFunctions.scala:1095)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
    at org.apache.spark.scheduler.Task.run(Task.scala:70)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
```

It looks like the fix is for the generated code to create a new ArrayList rather than use Arrays.asList, I was hoping to submit a PR but I couldn't figure out how to make this change. I'm happy to create a standalone project that demonstrates this issue if that would be helpful.

I'm using java 1.7, avro 1.7.7 and sbt-avrohugger 0.6.1.

**Comment by tackley** [comment]
I've created a standalone project that demonstrates this issue: https://github.com/tackley/avrohugger-list-issue

I realised I could manually edit the generated code to explore how to make it work, you can see my conclusion in the README.

**Comment by julianpeeters** [comment]
Thanks for uncovering this bug, and for what seems like will be a good fix!

> I was hoping to submit a PR

I'm happy to take a look and publish this weekend/early next week. [PutGenerator](https://github.com/julianpeeters/avrohugger/blob/scavro/avrohugger-core/src/main/scala/format/specific/methods/PutGenerator.scala) (and GetGenerator) needs to be updated, and then update a [unit test](https://github.com/julianpeeters/avrohugger/blob/master/avrohugger-core/src/test/scala/specific/SpecificGeneratorSpec.scala#L241) or two, and finally [integration tests](https://github.com/julianpeeters/avrohugger/tree/scavro/avrohugger-core/src/sbt-test/projects/SpecificSerializationTests) that use your example to test object reuse during actual de/serialization. And this should probably be done for the other datatypes as well. Worst, the integration tests have been piling up with lot of code-duplication, and it seems like it's time for a test revamp like [avro-macros](https://github.com/julianpeeters/avro-scala-macro-annotations/commit/517f680cd7a13b1ea0e5560f200e455f2ba47f4f) got recently. Of course I wouldn't mind the help but given the technical debt, and my recent experience fixing the avro-macro tests, I might suggest that I get avrohugger's tests updated first, and then I can walk you through if you'd like to become a contributor.

> but I couldn't figure out how to make this change

Aside from being a new project to you, and a confusing one since it uses compile-time meta-programming, by chance can you identify a sticking point that I might be able to make more clear?

**Comment by tackley** [comment]
Thanks. Everything was pretty easy to follow, it was just the compile-time meta-programming that got me - specifically understanding what the operators exposed by `TreehuggerDSLs` actually did. My technical language structure knowledge is not that high, so I suspect the issue is with me rather than your code :)

**Comment by julianpeeters** [comment]
Hi Graham, sorry for the delay on this.

I've updated the tests to demonstrate this issue. They are failing now, but should be fixable with a few changes.

If you'd still like to become a contributor, then fork this project, checkout the `list_issue` branch, and make the necessary changes to the field converters that GetGenerator and PutGenerator use so that `sbt test` succeeds. Specifically, [this line](https://github.com/julianpeeters/avrohugger/tree/list_issue/avrohugger-core/src/main/scala/format/specific/converters) needs a couple updates, and [this line](https://github.com/julianpeeters/avrohugger/blob/list_issue/avrohugger-core/src/main/scala/format/specific/converters/ScalaConverter.scala#L39) needs one too.

If you've moved on, no worries at all, I'll get to this myself If I don't hear back from you by next weekend.

**Comment by julianpeeters** [comment]
This should now be fixed in avrohugger version 0.8.2, sbt-avrohugger version 0.7.2
